keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -666,7 +773,1310 @@ ways.  One is that we are dispatching requests to be served rather
+regular dispatch logic does its thing.  When request `(i,j)` finishes
+execution and its actual duration is learned, the queue's
+`virtualStart` is adjusted by adding the product of `width(i,j)` and
+the correction to `len(i,j)`.
+
+That is the only adjustment made when the correct value of `len(i,j)`
+is learned.  This ignores other consequences that the equations above
+call for.  Changing `len(i,j)` can change `rho(i,t)` for some times
+`t`.  That can change `mu_equal(t)`, and thus the scheduling in all
+the queues.
+
+To advance the cursor, the implementation iterates through all the
+queues to find, among those that have requests waiting in the real
+world, the one with minimal next `F` (with tie broken appropriately).
+This costs `O(N)` compute time, where `N` is the number of queues.
+
+Using an equal allocation of concurrency rather than a max-min fair
+one is [issue
+95979](https://github.com/kubernetes/kubernetes/issues/95979).  In
+scenarios where the two allocations are different, equal allocation
+gives some queues more concurrency than they can use and gives other
+queues less than they should get.
+
+One consequence of this mis-allocation is that while a queue uses less
+than `mu_equal` but is non-empty at every moment when a request could
+be dispatched from it, the equations above say that the queue gets
+work done faster than it actually does.  The equations above assign
+`B` and `E` values that reflect an impossibly high rate of progress
+for such a queue.  That is, these values get progressively more early.
+The queue is effectively building up "credit" in artificially low `B`
+and `E` values, and can build up an arbitrary amount of this credit.
+Then an arbitrarily large amount of work could suddenly arrive to that
+queue and crowd out other queues for an arbitrarily long time.  To
+mitigate this problem, the implementation has a special step that
+effectively prevents `B` of the next request to dispatch from dropping
+below the current time.  But that solves only half of the problem.
+Other queueus may accumulate a corresponding deficit (inappropriately
+large values for `B` and `E`).  Such a queue can have an arbitrarily
+long burst of inappropriate lossage to other queues.
+
+### Fair Queuing for Server Requests, with max-min fairness and serial virtual execution
+
+In this quadrant we achieve max-min fairness and a non-empty queue has
+exactly one request executing at a given time in the virtual world.
+
+#### Fair Queuing for Server Requests, with max-min fairness and serial virtual execution, behavior
+
+The following behavior solves the problem.
+
+We imagine a virtual world in which request executions are scheduled
+(queued and executed) differently than in the real world.  In the
+virtual world, requests are generally executed with more or less
+concurrency than in the real world and with infinitesimal
+interleaving.
+
+The virtual world uses the same clock as the real world.
+
+Define the following.
```

Comment:
Do we need to copy all of that? You already defined all of that above...

Reply:
As mentioned in the meeting, my goal is to have each of the three quadrants stand on its own; I expect we will choose one and move the other two to the section about alternatives considered.

Reply:
SGTM